application/models/services/QuestionGroupService.php

```diff
@@ -0,0 +1,265 @@
+<?php
+
+namespace LimeSurvey\Models\Services;
+
+use LimeExpressionManager;
+use Survey;
+use Permission;
+use QuestionGroup;
+use QuestionGroupL10n;
+use Exception;
+use CException;
+use CHttpException;
+
+class QuestionGroupService
+{
+    private Permission $modelPermission;
+    private Survey $modelSurvey;
+
+    /**
+     *
+     */
+    public function __construct(
+        Permission $modelPermission,
+        Survey $modelSurvey
+    ) {
+        $this->modelPermission = $modelPermission;
+        $this->modelSurvey = $modelSurvey;
+    }
+
+    /**
+     * Updates a question group and all the languages.
+     *
+     * @param $surveyId int the survey id
+     * @param $questionGroupID int the question group id
+     * @param $input  array   has the data for a question group, including an array for languages
+     *                     ['questionGroup']
+     *                          [gid]
+     *                          [sid]
+     *                          [group_order]
+     *                          [randomization_group]
+     *                          [grelevance]
+     *                     ['questionGroupI10N']
+     *                         [en]
+     *                            [group_name]
+     *                            [description]
+     *                         [...]    //more languages
+     * @return QuestionGroup
+     */
+    public function updateGroup($surveyId, $questionGroupID, $input)
+    {
+        $survey = $this->getSurvey($surveyId);
```

Comment:
@olleharstedt Often the service logic needs to load multiple objects from persistence in order to carry out its work. The objects to be loaded may even depend on some initial logic to work out which objects we are need to work on.

This logic does not belong in the controller. 

There may be an argument for organising the business logic in order to isolate this work but the refactoring we are doing right now is minimal work in order to achieve two things:

1. Reusability
2. Testability

Anything refactoring beyond that is out of scope.

Reply:
In this case, the loading of the survey does not depend on any logic tho. So the controller code could easily be:

```php
$service->updateGroup(Survey::model()->findByPk($surveyId), $questionGroupId, $input)
```

Function composition, pretty basic. :) But I've made this argument before, and you didn't agree then either.

Reply:
Sure in this case its relatively simply in other cases its not. Regardless, in all cases, this logic for loading objects does not belong in the controller.

Reply:
I disagree wholeheartedly. :) In this case, it very much belongs to the _setup_ phase of the service, which is in the controller.

Reply:
@olleharstedt 

The point of the application model (which are implementing as services) is that it abstracts away the business logic of the application. It is a context agonistic (http, cli, unit test) boundary point. It allows us to interact with the application via an API in the same way that we might consume any other API. 

When designing an API the main focus is the interface. How does the consumer interact with the API? The ideal implementation will keep the API as simple as possible. The less the consumer needs to know about the internal workings of the API the better. 

When you interact with the business model you should not need to be aware of any implementation details. You should not need to know about HOW data is persisted in order to interact with it. The implementation details of data persistence is handled by the model. 

As I have said, there is an argument for different approaches to separating persistence logic from pure business logic. Regardless of how you handle that separation, it should still be handled within model, since its not a concern of the consumer of the model. The consumer just wants to add a question or update a survey setting. They don't care HOW you do it.

Yes I agree is there code smell, but thats not an argument for putting persistence logic in the controller and duplicating it in unit tests and cli scripts and anywhere else we want to use the model.

Reply:
@kevin-foster-uk The controller is the _correct_ place to have persistence logic, since it's the "imperative shell" of the application. :D Service classes should have as few side-effects as possible, in my opinion - that makes them way easier to test and reuse in different contexts.

Your other arguments seem to agree with the point of lifting loading data out of the service class when possible, since it makes the service class _more_ agnostic of how it's loaded.

Duplicating the loading in unit test is not needed - it won't be loaded at all, it will just be a mock. Instead of mocking the loading, you mock the survey object directly. No loading needed.

But yes, some duplication will appear if you use the service class from CLI or REST. That's totally fine, the same type of duplication happens already with the DI setup, right?

Reply:
@olleharstedt We are not designing an application from scratch here. We have some existing code, we are trying to make it reusable and testable. We can argue as much as you want about how it should be organised right now its irrelevant. This is what we have.

If you want to take over directing the refactoring tasks and take it in a new direction we can discuss that.

Reply:
Just a simple suggestion to improve testability of the service class, Kevin. :) We can mark it as resolved if we can't agree.